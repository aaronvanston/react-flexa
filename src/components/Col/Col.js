import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { themeProvider } from '../../theme';
import { filterProps, mediaQuery, columnWidth, gutter, CSSProperty } from '../../helpers';

const Col = styled(props =>
  React.createElement(props.elementType, filterProps(props, Col.propTypes)),
)`
  // Initial component properties
  box-sizing: border-box;

  ${props => themeProvider.breakpointsKeys(props).map(breakpoint => mediaQuery(props)[breakpoint]`
    // Generated Display
    ${CSSProperty(props, breakpoint, 'display')}

    // Generate gutter
    ${gutter.col(props, breakpoint)}

    // Generate flex rule before width, this avoid override
    ${CSSProperty(props, breakpoint, 'flex')}

    // Generate column width
    ${columnWidth(props, breakpoint)}

    // Responsive Flexbox properties
    ${CSSProperty(props, breakpoint, 'order')}
    ${CSSProperty(props, breakpoint, 'align-self')}
  `)}
`;

Col.defaultProps = {
  order: 0,
  alignSelf: 'auto',
  elementType: 'div',
  display: 'block',
  flex: '0 0 auto',
};

const displayOptions = ['flex', 'inline-flex', 'block', 'none', 'inline-block'];
export const alignSelfOptions = ['auto', 'flex-start', 'flex-end', 'center', 'baseline', 'stretch'];

Col.propTypes = {
  xs: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.oneOf(['hidden', 'auto'])]),
  sm: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.oneOf(['hidden', 'auto'])]),
  md: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.oneOf(['hidden', 'auto'])]),
  lg: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.oneOf(['hidden', 'auto'])]),

  gutter: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.shape({
      xs: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
      ]),
      sm: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
      ]),
      md: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
      ]),
      lg: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
      ]),
    }),
  ]),

  order: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.shape({
      xs: PropTypes.number,
      sm: PropTypes.number,
      md: PropTypes.number,
      lg: PropTypes.number,
    }),
  ]),

  alignSelf: PropTypes.oneOfType([
    PropTypes.oneOf(alignSelfOptions),
    PropTypes.shape({
      xs: PropTypes.oneOf(alignSelfOptions),
      sm: PropTypes.oneOf(alignSelfOptions),
      md: PropTypes.oneOf(alignSelfOptions),
      lg: PropTypes.oneOf(alignSelfOptions),
    }),
  ]),

  elementType: PropTypes.string,

  flex: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]),

  display: PropTypes.oneOfType([
    PropTypes.oneOf(displayOptions),
    PropTypes.shape({
      xs: PropTypes.oneOf(displayOptions),
      sm: PropTypes.oneOf(displayOptions),
      md: PropTypes.oneOf(displayOptions),
      lg: PropTypes.oneOf(displayOptions),
    }),
  ]),

  children: PropTypes.node,
};

Col.displayName = 'Col';

export default Col;
