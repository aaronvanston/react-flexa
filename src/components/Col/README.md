# Col

A component for creating a column containing other elements, within a grid system. A `<Col/>` must be wrapped in a parent `<Row/>`.

## Properties

View detailed property information in [Col props](https://github.com/aaronvanston/react-flexa#col-props) within the react-flexa readme.

- `xs`:
  - **Number:** Sets the width of the col based the column configuration (12 by default) for the XS breakpoint and up.
  - **String:** Set a width like string `xs="390px"` to fixed have a fixed colum, or create a calc `xs="calc(100% - 390px)"` base on sibling to have a responsive column.
- `sm`:
  - **Number:** Sets the width of the col based the column configuration (12 by default) for the SM breakpoint and up.
  - **String:** Set a width like string `sm="390px"` to fixed have a fixed colum, or create a calc `sm="calc(100% - 390px)"` base on sibling to have a responsive column.
- `md`:
  - **Number:** Sets the width of the col based the column configuration (12 by default) for the MD breakpoint and up.
  - **String:** Set a width like string `md="390px"` to fixed have a fixed colum, or create a calc `md="calc(100% - 390px)"` base on sibling to have a responsive column.
- `lg`:
  - **Number:** Sets the width of the col based the column configuration (12 by default) for the LG breakpoint and up.
  - **String:** Set a width like string `lg="390px"` to fixed have a fixed colum, or create a calc `lg="calc(100% - 390px)"` base on sibling to have a responsive column.
- `gutter`: Sets the width of the gutter to be used between columns.
- `order`: Controls the order in which they appear in the flex container.
- `alignSelf`: Controls vertical alignment in the flex container.
- `display`: This defines the display property of the col, inline, block, flex etc.

## Usage

### Basic

```js
import React from 'react';
import { Row, Col } from 'react-flexa';

export default () => (
  <Row>
    <Col xs={6}></Col>
    <Col xs={6}></Col>
  </Row>
);
```

### Responsive

```js
import React from 'react';
import { Row, Col } from 'react-flexa';

export default () => (
  <Row>
    <Col xs={12} md={6} lg={4} order={{ xs: 2, lg: 1 }}></Col>
    <Col xs={12} md={6} lg={8} order={{ xs: 1, lg: 2 }}></Col>
  </Row>
);
```

### One fixed column

```js
import React from 'react';
import { Row, Col } from 'react-flexa';

export default () => (
  <Row>
    <Col xs={12} sm={6} md={6} lg="390px"></Col>
    <Col xs={12} sm={6} md={6} lg="calc(100% - 390px)"></Col>
  </Row>
);
```

```js
import React from 'react';
import { Row, Col } from 'react-flexa';

export default () => (
  <Row>
    <Col xs={12} sm={6} md={6} lg="390px"></Col>
    <Col xs={12} sm={6} md={6} lg="calc(80% - 390px)"></Col>
    <Col xs={12} sm={6} md={6} lg="20%"></Col>
  </Row>
);
```

Read more about react-flexa [responsive api using objects.](https://github.com/aaronvanston/react-flexa#responsive-api-using-objects)
