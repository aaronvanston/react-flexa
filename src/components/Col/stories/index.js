export Col from './Col';
export Fluid from './Fluid';
export Responsive from './Responsive';
export AutoGrow from './AutoGrow';
export Fixed from './Fixed';
