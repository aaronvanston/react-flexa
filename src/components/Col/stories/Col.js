/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { select, number, object } from '@storybook/addon-knobs';

import { Box } from '../../../../stories/components';
import { Row, Col } from '../../';
import { alignSelfOptions } from '../Col';

export default () => (
  <Row>
    <Col xs={12} sm={4} md={6} lg={3}>
      <Box>col</Box>
    </Col>

    <Col
      xs={number('xs', 12)}
      sm={number('sm', 4)}
      md={number('md', 6)}
      lg={number('lg', 3)}
      gutter={object('gutter (rem)', { xs: 0.5, sm: 0.5, md: 1, lg: 1 })}
      order={number('order', 0)}
      alignSelf={select('align-self', alignSelfOptions, 'auto')}
    >
      <Box bgColor="#ff9800">customize me!</Box>
    </Col>

    <Col xs={12} sm={4} md={6} lg={3}>
      <Box>col</Box>
    </Col>

    <Col xs={12} sm={4} md={6} lg={3}>
      <Box>col</Box>
    </Col>
  </Row>
);
