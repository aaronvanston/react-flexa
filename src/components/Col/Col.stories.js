/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs } from '@storybook/addon-knobs';
import withReadme from 'storybook-readme/with-readme';

import { Col, Fluid, Responsive, AutoGrow, Fixed } from './stories';
import README from './README.md';

storiesOf('Col', module)
  .addDecorator(withKnobs)
  .addDecorator(withReadme(README))
  .add('Default', () => <Col />)
  .add('Fluid example', () => <Fluid />)
  .add('Responsive example', () => <Responsive />)
  .add('Auto grow example', () => <AutoGrow />)
  .add('One fixed col', () => <Fixed />);
