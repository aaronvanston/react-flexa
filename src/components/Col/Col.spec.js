import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import Col from './Col';

describe('style rendering', () => {
  test('renders corrects', () => {
    const wrapper = mount(<Col />);
    expect(wrapper).toMatchSnapshot();
  });

  test('generates default css', () => {
    const tree = renderer.create(<Col />).toJSON();
    expect(tree).toHaveStyleRule('box-sizing', 'border-box');
    expect(tree).toHaveStyleRule('display', 'block');
    expect(tree).toHaveStyleRule('flex', '0 0 auto');
    expect(tree).toHaveStyleRule('align-self', 'auto');
    expect(tree).toHaveStyleRule('padding-left', 'calc(0.5rem / 2)');
    expect(tree).toHaveStyleRule('padding-right', 'calc(0.5rem / 2)');
  });

  test('generates custom css', () => {
    const tree = renderer.create(<Col order={2} alignSelf="flex-start" />).toJSON();
    expect(tree).toHaveStyleRule('order', '2');
    expect(tree).toHaveStyleRule('align-self', 'flex-start');
  });
});

describe('Create element', () => {
  test('Should by default create a div element', () => {
    const wrapper = shallow(<Col />);
    expect(wrapper.dive().type()).toEqual('div');
  });

  test('Should create a custom element if provided', () => {
    const wrapper = shallow(<Col elementType="span" />);
    expect(wrapper.dive().type()).toEqual('span');
    expect(wrapper.dive().type()).not.toEqual('div');
  });

  test('generates custom with custom display css property', () => {
    const tree = renderer.create(<Col display={'flex'} />).toJSON();
    expect(tree).toHaveStyleRule('display', 'flex');
  });
});
