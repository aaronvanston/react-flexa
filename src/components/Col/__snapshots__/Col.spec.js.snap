// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`style rendering renders corrects 1`] = `
<Col
  alignSelf="auto"
  autoGrow={false}
  elementType="div"
  flex={false}
  order={0}
>
  <Component
    alignSelf="auto"
    autoGrow={false}
    className="sc-bdVaJa jURNTC"
    elementType="div"
    flex={false}
    order={0}
  >
    <div
      className="sc-bdVaJa jURNTC"
    />
  </Component>
</Col>
`;
